### PR TITLE
Implement due dates for cards

### PR DESCRIPTION
**Pull request by juliushaertl**
Fix for #106 

cc @artemanufrij @pixelipo

### CODE REVIEW

**Comment by artemanufrij** [comment]
screenshot_20170616_182712

File: img/calendar.svg [hunk 1]
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" height="32" width="32" viewbox="0 0 32 32"><path d="M8 2c-1.108 0-2 .892-2 2v4c0 1.108.892 2 2 2s2-.892 2-2V4c0-1.108-.892-2-2-2zm16 0c-1.108 0-2 .892-2 2v4c0 1.108.892 2 2 2s2-.892 2-2V4c0-1.108-.892-2-2-2zM11 6v2c0 1.662-1.338 3-3 3S5 9.662 5 8V6.125A3.993 3.993 0 0 0 2 10v16c0 2.216 1.784 4 4 4h20c2.216 0 4-1.784 4-4V10a3.993 3.993 0 0 0-3-3.875V8c0 1.662-1.338 3-3 3s-3-1.338-3-3V6zM6.094 16h19.812a.09.09 0 0 1 .094.094v9.812a.09.09 0 0 1-.094.094H6.094A.09.09 0 0 1 6 25.906v-9.812A.09.09 0 0 1 6.094 16z"/></svg>
```
Comment (pixelipo) [inline img/calendar.svg]: This file is available in core/img/places

File: img/calendar-white.svg [hunk 2]
```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (pixelipo) [inline img/calendar-white.svg]: I've created a PR to add white calendar icon to core in nextcloud/server#5447 - so this should become redundant

File: css/style.css [hunk 3]
```diff
@@ -458,14 +458,49 @@ button.button-inline:hover {
...
+	opacity: 0.7;
+	margin-top:2px;
+}
+.overdue {
+	background-color: #e12419;
+	color: #fff;
+}
+.due.now {
+	background-color: #fbd850;
+	color: #000;
```
Comment (pixelipo) [inline css/style.css]: not needed - black is default

File: css/style.css [hunk 4]
```diff
@@ -458,14 +458,49 @@ button.button-inline:hover {
...
+}
+.due.next {
+	background-color: #22ac2a;
+	color: #fff;
 }
 
+.due .icon-calendar {
```
Comment (pixelipo) [inline css/style.css]: you should move this over to `.due .icon` as it is a default setting. Other `.something .icon-calendar` declarations will override background-image and margin-right is needed for all.

File: css/style.css [hunk 5]
```diff
@@ -503,9 +538,9 @@ button.button-inline:hover {
 	margin-bottom: 10px;
 }
 
-#card-dates span {
+#card-meta #duedate {
```
Comment (pixelipo) [inline css/style.css]: it's usually not a good idea to used ID as a CSS selector. I would recommend adding a class and applying style to it.

File: templates/main.php [hunk 6]
```diff
@@ -36,16 +36,17 @@
 Util::addScript('deck', 'vendor/markdown-it/dist/markdown-it.min');
 Util::addScript('deck', 'vendor/angular-markdown-it/dist/ng-markdownit.min');
 Util::addScript('deck', 'vendor/markdown-it-link-target/dist/markdown-it-link-target.min');
+Util::addScript('deck', 'vendor/jquery-timepicker/jquery.ui.timepicker');
 
-if(!\OC::$server->getConfig()->getSystemValue('debug', false)) {
+if(true && !\OC::$server->getConfig()->getSystemValue('debug', false)) {
```
Comment (juliushaertl) [inline templates/main.php]: :see_no_evil: Some debugging statement I forgot about.

File: js/filters/dateFilters.js [hunk 7]
```diff
@@ -20,8 +20,42 @@
...
+app.filter('dateToTimestamp', function() {
+	return function (date) {
+		return Date.parse(date);
+	}
+});
+
+app.filter('parseDate', function() {
+	return function (date) {
+		if(moment(date).isValid()) {
+			return moment(date).format('YYYY-MM-DD');
```
Comment (juliushaertl) [inline js/filters/dateFilters.js]: Yes, but that is out of scope for now, since we want to release 0.2 soon. We should also be timezone aware. But let's address that afterwords.

File: templates/part.card.php [hunk 8]
```diff
@@ -45,6 +45,13 @@
 					style="background-color:#{{label.color}}; color:{{ label.color|textColorFilter }};">{{label.title}}</span>
 			</ui-select-choices>
 		</ui-select>
+
+		<div id="duedate">
+			<input class="datepicker-input medium focus" type="text" placeholder="Set a due date" value="{{ cardservice.getCurrent().duedate | parseDate }}" datepicker="due" />
+			<input class="timepicker-input medium focus" type="text" placeholder="00:00:00" ng-if="cardservice.getCurrent().duedate" value="{{ cardservice.getCurrent().duedate | parseTime }}" timepicker="due" />
+			<button class="icon icon-delete button-inline" title="Remove duedate" ng-if="cardservice.getCurrent().duedate" ng-click="resetDuedate()"></button>
```
Comment (juliushaertl) [inline templates/part.card.php]: :+1:

File: img/calendar-white.svg (hunk 2 above)
Comment (juliushaertl) [inline img/calendar-white.svg]: :+1: But we should keep it for now until all supported Nextcloud versions ship with that icon.

**Comment by juliushaertl** [comment]
@pixelipo @artemanufrij Thanks for the comments, should be all fixed now.

File: img/calendar-white.svg (hunk 2 above)
Comment (pixelipo) [inline img/calendar-white.svg]: I agree, but please use the optimized version - either copy from my PR or run scour command:
`scour -i calendar-white.svg -o calendar-white-new.svg --create-groups --enable-id-stripping --enable-comment-stripping --shorten-ids --remove-metadata --strip-xml-prolog --no-line-breaks`

Comment (juliushaertl) [inline img/calendar-white.svg]: done :)

**Comment by artemanufrij** [comment]
margin is still a little bit inexact:
spectacle qv4727

I think the icons should be adjusted

**Comment by juliushaertl** [comment]
Thanks @artemanufrij Let's merge this. :rocket:

**Comment by Hanzei** [comment]
Nice work! :tada:
